pkg/statistics/handle/autoanalyze/priorityqueue/queue_v2.go

```diff
@@ -0,0 +1,247 @@
+
+package priorityqueue
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/ddl/notifier"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/sessionctx/variable"
+	"github.com/pingcap/tidb/pkg/statistics/handle/autoanalyze/exec"
+	"github.com/pingcap/tidb/pkg/statistics/handle/autoanalyze/internal/heap"
+	statslogutil "github.com/pingcap/tidb/pkg/statistics/handle/logutil"
+	statstypes "github.com/pingcap/tidb/pkg/statistics/handle/types"
+	statsutil "github.com/pingcap/tidb/pkg/statistics/handle/util"
+	"github.com/pingcap/tidb/pkg/util"
+	"go.uber.org/zap"
+)
+
+const (
+	lastAnalysisDurationRefreshInterval = time.Minute * 10
+)
+
+// AnalysisPriorityQueueV2 is a priority queue for TableAnalysisJobs.
+type AnalysisPriorityQueueV2 struct {
+	inner                  *heap.Heap[int64, AnalysisJob]
+	statsHandle            statstypes.StatsHandle
+	calculator             *PriorityCalculator
+	autoAnalysisTimeWindow atomic.Value // stores *autoAnalysisTimeWindow
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     util.WaitGroupWrapper
+
+	// initialized is a flag to check if the queue is initialized.
+	initialized atomic.Bool
+}
+
+// NewAnalysisPriorityQueueV2 creates a new AnalysisPriorityQueue2.
+func NewAnalysisPriorityQueueV2(handle statstypes.StatsHandle) *AnalysisPriorityQueueV2 {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	return &AnalysisPriorityQueueV2{
+		statsHandle: handle,
+		calculator:  NewPriorityCalculator(),
+		ctx:         ctx,
+		cancel:      cancel,
+	}
+}
+
+// IsInitialized checks if the priority queue is initialized.
+func (pq *AnalysisPriorityQueueV2) IsInitialized() bool {
+	return pq.initialized.Load()
+}
+
+// Initialize initializes the priority queue.
+func (pq *AnalysisPriorityQueueV2) Initialize() error {
+	if pq.initialized.Load() {
+		statslogutil.StatsLogger().Warn("Priority queue already initialized")
+		return nil
+	}
+
+	start := time.Now()
+	defer func() {
+		statslogutil.StatsLogger().Info("Priority queue initialized", zap.Duration("duration", time.Since(start)))
+	}()
+
+	keyFunc := func(job AnalysisJob) (int64, error) {
+		return job.GetTableID(), nil
+	}
+	// We want the job with the highest weight to be at the top of the priority queue.
+	lessFunc := func(a, b AnalysisJob) bool {
+		return a.GetWeight() > b.GetWeight()
+	}
+	pq.inner = heap.NewHeap(keyFunc, lessFunc)
+	if err := pq.init(); err != nil {
+		pq.Close()
+		return errors.Trace(err)
+	}
+
+	// Start a goroutine to maintain the priority queue.
+	pq.wg.Run(pq.run)
+	pq.initialized.Store(true)
+	return nil
+}
+
+// init initializes the priority queue and adds jobs to it.
+func (pq *AnalysisPriorityQueueV2) init() error {
+	return statsutil.CallWithSCtx(pq.statsHandle.SPool(), func(sctx sessionctx.Context) error {
+		parameters := exec.GetAutoAnalyzeParameters(sctx)
+		err := pq.setAutoAnalysisTimeWindow(parameters)
```

Comment:
Oh, here i raise a question, when we refresh the window if we change the global value?
I searched the usage and seems that we haven't handled it yet.

Reply:
We will do it here https://github.com/pingcap/tidb/pull/55889/files#r1777880034.
But I guess we can do it in both processes.

Reply:
Updated.